proxy: support for public unauthenticated routes

nareddyt opened the pull request:
Fixes #129

### Implementation Description

- Implemented a new access control option that allows public, unauthenticated access to specific targets.
- Added bypass check to `Proxy::shouldSkipAuthentication`

### Testing Done

- Ensured all unit tests pass in `proxy/`
- Ran the following manual integration test

#### Policy Configuration

```yaml
- from: hello1.homelab.tejunareddy.com
  to: http://localhost:8081
  allowed_users:
    - tejunareddy@gmail.com
- from: hello2.homelab.tejunareddy.com
  to: http://localhost:8082
  allowed_users:
    - not_a_user@yahoo.com
- from: hello3.homelab.tejunareddy.com
  to: http://localhost:8083
  allow_public_unauthenticated_access: true
```

#### Steps

1. Sign out of pomerium with `/.pomerium/signout`
2. Ensure 200 OK for hello3.homelab.tejunareddy.com, but 403 Forbidden for hello1 and hello2.
3. Sign into tejunareddy@gmail.com
4. Ensure 400 OK for hello3 and hello1, but 403 Forbidden for hello2.

### Other Notes

In order for this to work correctly, #132 must be fixed. I locally used the code in #133 to fix this issue, and tested this PR on top of that.

**Checklist**:
- [x] updated docs
- [x] unit tests added
- [x] related issues referenced
- [x] ready for review

nareddyt commented on docs/reference/readme.md:
```diff
@@ -185,6 +185,19 @@ Allowed domains is a collection of whitelisted domains to authorize for a given
 
+#### Public Access
+
+- `yaml`/`json` setting: `allow_public_unauthenticated_access`
```
Let me know if this name is OK, or if you'd prefer something along the lines of `disable_all_access_control`

desimone replied:
I can't think of anything better. `disable_all_access_control` **is** descriptive. 

I'll try to see what other products call their "turn-off-all-access-control" flags just to see but I think this is fine.

desimone replied:
You pick, I'm open to either. 👍

nareddyt commented on proxy/handlers.go:
```diff
@@ -185,6 +185,18 @@ func (p *Proxy) shouldSkipAuthentication(r *http.Request) bool {
+		}
+
+		log.FromRequest(r).Error().Msg("proxy: route marked as public contains whitelists, " +
+			"disabling public access and using whitelists")
```
Alternatively, this check can be placed in [internal\policy\policy.go::validate](https://github.com/pomerium/pomerium/blob/409f44a9be1d8bbdf768df2b508b7d427d279f72/internal/policy/policy.go#L37). Would it make more sense to do this validation when the server is started and just completely disallow both options from being enabled? I admit it does seem weird running this check in the dataplane when we could have ran it at startup in the controlplane

desimone replied:
Yeah, I think it would more sense we should pop an error when loading the policy instead. Saves resources, and avoids any user confusion.

Plus, I was thinking that at some point it would be nice to be able to abstract out a the policy validate logic as a stand-alone linter so admins can check their policy prior to any updates.

desimone commented on docs/reference/readme.md:
```diff
@@ -185,6 +185,19 @@ Allowed domains is a collection of whitelisted domains to authorize for a given
+- Optional
+- Default: `false`
+
+**Use with caution:** Allow all requests for a given route, bypassing authentication. 
```
Sort of obvious but (how do you authorize without authenticating first), but we might as well be explicit that this bypasses all authentication **and** authorization.

nareddyt commented:
Moved the validation from the data plane to the control plane. The following error gets printed when a policy contains both a whitelist and public access:

```
FTL cmd/pomerium: authorize error="authorize: `POLICY_FILE` route at index 2: route marked as public but contains whitelists"
```

Additional changes:

- Updated docs
- Updated unit tests
- Ran manual integration tests (described above)

Still haven't decided on the name, let me know if that needs to be changed to `disable_all_access_control` :)

desimone commented on internal/policy/policy.go:
```diff
@@ -39,10 +43,17 @@ func (p *Policy) validate() (err error) {
 	}
+
+	// Only allow public access if no other whitelists are in place
+	if p.AllowPublicUnauthenticatedAccess && (p.AllowedDomains != nil || p.AllowedGroups != nil || p.AllowedEmails != nil) {
```
Could we add a test or two for this?

nareddyt replied:
I added two tests to `proxy/handlers_test.go` that should test for this. Code coverage report shows this branch is tested. Is there a more in-depth test you'd like to see?

desimone replied:
Maybe I've got this wrong but this but it shows as uncovered for me (this would be for policy, not handlers right?)

https://codecov.io/gh/pomerium/pomerium/compare/c53c07c274f23d97f24e56ab1e75e8a5a288b92c...0952a1de46ba5faf1828567c228d536c8a810775/src/internal/policy/policy.go#L54

nareddyt replied:
Sorry i meant `proxy/proxy_test.go`

desimone replied:
Ah got it, yeah it's weird that codecov isn't showing this as covered as 

```go
{"policy public and whitelist bad", publicWithWhitelistPolicy, true},
``` 

should be doing this right?

nareddyt replied:
Yeah, that line should be checking that. That's really weird. I see that branch covered after running the `TestOptions_Validate` function locally.

![image](https://user-images.githubusercontent.com/11142171/58217587-6837b080-7cd1-11e9-8a3c-c7b0cb3686c5.png)

desimone replied:
LGTM then 👍

nareddyt commented:
By the way, I'm having a lot of trouble squashing my commits. It might be because I merged `master` into the branch a few times. My bad on that, I'll remember to always rebase onto master for any future contributions.

Do you want to try squashing them? I tried and pushed the squashed commit to [this branch on my fork](https://github.com/nareddyt/pomerium/tree/unauthenticated-routes-squash), but it shows the diff for all the merges to master as well...

desimone commented:
I can squash your commits no problem, the only issue is the git commit history will look like I made the commits and not you. 

Does `git rebase -i master` work?

nareddyt commented:
Fixed! I did do `git rebase -i master`, but I forgot to fetch from upstream into my `master` branch. The diff looks good now

desimone commented:
🎉 thank you for the contribution!
